src/onnx/parse_einsum.cpp

```diff
@@ -0,0 +1,766 @@
+                          const std::vector<instruction_ref>& args) const
+    {
+        if(not contains(info.attributes, "equation"))
+            MIGRAPHX_THROW("Equation attribute is required");
+        std::string equation = info.attributes.at("equation").s();
+
+        const auto [terms, unique_labels, ellipses_ndim] = analyze_equation(equation, args);
+        const auto map_mat    = make_mapping_matrix(terms, unique_labels, ellipses_ndim);
+        const auto duplicates = find_duplicates(terms);
+
+        // Holds the mapping matrix representations of the two terms being processed
+        // cur_pair[0] acts as the accumulator for previously processed inputs
+        // cur_pair[1] holds the representation for the current input
+        // As operations are added to the einsum graph, cur_pair gets manipulated
+        int_mat cur_pair = make_matrix(2, map_mat[0].size(), -1);
+
+        instruction_ref cur_op;
+        std::optional<instruction_ref> last_op;
+        // Perform a left fold on the inputs
+        for(auto arg_idx = 0; arg_idx < args.size(); ++arg_idx)
+        {
+            cur_op      = args[arg_idx];
+            cur_pair[1] = map_mat[arg_idx];
+
+            cur_op =
+                preprocess_input(info, cur_op, duplicates[arg_idx], map_mat, arg_idx, cur_pair);
+
+            if(last_op)
+                cur_op = process_pair(info, *last_op, cur_op, map_mat, arg_idx, cur_pair);
+
+            last_op     = cur_op;
+            cur_pair[0] = cur_pair[1];
+        }
+
+        return finalize_output(info, cur_op, map_mat, cur_pair);
+    }
+
+    // Equation Parsing
+
+    std::tuple<string_vec, std::string, size_t>
+    analyze_equation(std::string_view equation, const std::vector<instruction_ref>& args) const
+    {
+        std::tuple<string_vec, std::string, size_t> ret;
+        auto& [terms, unique_labels, ellipses_ndim] = ret;
+
+        auto [input_terms, output_term, label_count, explicit_form] = parse_equation(equation);
+
+        ellipses_ndim = validate_input_terms(input_terms, args);
+        if(not output_term.empty())
+            validate_output_term(output_term, label_count, ellipses_ndim);
+        else if(not explicit_form)
+            output_term = generate_output_term(label_count, ellipses_ndim);
+
+        terms = std::move(input_terms);
+        terms.emplace_back(std::move(output_term));
+        for(const auto [l, _] : label_count)
+            unique_labels += l;
+
+        return ret;
+    }
+
+    std::tuple<string_vec, std::string, char_int_map, bool>
+    parse_equation(std::string_view equation) const
+    {
+        std::tuple<string_vec, std::string, char_int_map, bool> ret;
+        // cppcheck-suppress variableScope
+        auto& [input_terms, output_term, label_count, explicit_form] = ret;
+
+        std::string term;
+        bool has_ellipsis = false;
+        explicit_form     = false;
+
+        for(int i = 0; i < equation.size(); ++i)
```

Comment:
I think we should do lexing to parse the equation. We have some lexing functions here for the json:

https://github.com/ROCm/AMDMIGraphX/blob/develop/src/convert_to_json.cpp#L58

We could move this to a header file so we can reuse that to lex the einsum. We could also refactor it to return `std::vector<string_view>` instead of using the `token` typedef. A `lex_equal` function could be added to do comparisons:

```cpp
auto lex_equal(std::string s)
{
    return [=](const char* start, const char* last) {
        auto n = last - start;
        if (n < s.size())
            return start;
        if (std::equal(start, start+s.size(), s))
            return start+s.size();
        return start;
    };
}

auto einsum_tokenize(const std::string& s)
{
    std::vector<lexer> lexers;

    lexers.push_back(lex_while(&isspace));
    lexers.push_back(lex_while(&isalpha));
    lexers.push_back(lex_equal("->"));
    lexers.push_back(lex_equal("..."));
    lexers.push_back(lex_equal(","));

    return tokenize(s.data(), s.data() + s.length(), lexers);
}
```

This should make it simpler to loop over the terms.

Reply:
Done.